src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -36,13 +36,190 @@ public class ProjectDependencyGraph
+            }
+
+            var newDependencySets = _lazyDependencySets;
+
+            if (!newDependencySets.IsDefault)
+            {
+                var builder = newDependencySets.ToBuilder();
+
+                foreach (var projectId in projectIds)
+                {
+                    builder.Add(ImmutableArray.Create(projectId));
+                }
+
+                newDependencySets = builder.ToImmutable();
+            }
+
+            return new ProjectDependencyGraph(
+                _projectIds.AddRange(projectIds),
+                referencesMap: _referencesMap,
+                reverseReferencesMap: _lazyReverseReferencesMap,
+                transitiveReferencesMap: _transitiveReferencesMap,
+                reverseTransitiveReferencesMap: _reverseTransitiveReferencesMap,
+                topologicallySortedProjects: newTopologicallySortedProjects,
+                dependencySets: newDependencySets);
+        }
+
+        internal ProjectDependencyGraph WithAdditionalProjectReferences(ProjectId projectId, IEnumerable<ProjectId> referencedProjectIds)
+        {
+            // Ensure we only go over the list once, since we'll be enumerating it many times. This also gives us an easy bail if we can no-op
+            referencedProjectIds = referencedProjectIds.ToList();
+
+            if (!referencedProjectIds.Any())
+            {
+                return this;
+            }
+
+            // Step #1: recompute our references map
+            var newReferencesMap = _referencesMap;
+
+            if (newReferencesMap.TryGetValue(projectId, out var existingReferences))
+            {
+                newReferencesMap = newReferencesMap.SetItem(projectId, existingReferences.Union(referencedProjectIds));
+            }
+            else
+            {
+                newReferencesMap = newReferencesMap.SetItem(projectId, referencedProjectIds.ToImmutableHashSet());
+            }
+
+            // Step #2: compute our reverse map, if we had one previously to build off of
+            var newReverseReferencesMap = _lazyReverseReferencesMap;
+
+            if (newReverseReferencesMap != null)
+            {
+                var builder = newReverseReferencesMap.ToBuilder();
+
+                foreach (var referencedProject in referencedProjectIds)
+                {
+                    if (builder.TryGetValue(referencedProject, out var reverseReferences))
+                    {
+                        builder[referencedProject] = reverseReferences.Add(projectId);
+                    }
+                    else
+                    {
+                        builder[referencedProject] = ImmutableHashSet.Create(projectId);
+                    }
+                }
+
+                newReverseReferencesMap = builder.ToImmutable();
+            }
+
+            // Step #3: update our forward transitive map, by combining the transitive references of the new projects into the transitive references of projectId.
```

Comment:
See the discussion in the main PR description about this part. I wrote steps 1 through 4 before doing a profile, when perhaps I should have written steps 1 and 2, and then remeasured to see if 3 and 4 were needed. (I got too carried away having fun with graphs!) On one hand this (assuming no bugs) should do a great job keeping things incrementally updated if we're able to cheaply do so, it does mean that we're holding onto memory for these transitive maps that previously we would have thrown away. Steps 1 and 2 should be kept because those were always being realized no matter what; 3 and 4 we _may_ be better off deleting.